lib/elixir/lib/calendar.ex

```diff
@@ -796,6 +710,75 @@ defmodule NaiveDateTime do
   end
 
   @doc """
+  Adds a specified amount of time to a `NaiveDateTime`.
+
+  Accepts an `integer` in any `unit` available from `System.time_unit`.
+  Negative values will be move backwards in time.
+
+  ## Examples
+
+      # adds seconds by default
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], 2)
+      ~N[2014-10-02 00:29:12]
+      # accepts negative offsets
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], -2)
+      ~N[2014-10-02 00:29:08]
+      # can work with other units
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], 2_000, :milliseconds)
+      ~N[2014-10-02 00:29:12]
+      # doesn't lose precision
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10.021], 21, :milliseconds)
+      ~N[2014-10-02 00:29:10.042]
+      # from gregorian seconds
+      iex> NaiveDateTime.add(~N[0000-01-01 00:00:00], 63579428950)
+      ~N[2014-10-02 00:29:10]
+  """
+  @spec add(NaiveDateTime.t, integer, System.time_unit) :: NaiveDateTime.t
+  def add(%NaiveDateTime{} = naive_datetime,
+          integer,
+          unit \\ :seconds) when is_integer(integer) do
+    ndt_microseconds = to_microseconds(naive_datetime)
+    added_microseconds = System.convert_time_unit(integer, unit, :microseconds)
+    sum = ndt_microseconds + added_microseconds
+
+    microsecond = rem(sum, 1_000_000)
+    precision =
+      if microsecond == 0, do: 0, else: Calendar.ISO.precision_for_unit(unit)
```

Comment:
Even if microsecond is zero, it may still have high precision!

For example: `~N[2014-10-02 00:29:10.000000]` is not the same as `~N[2014-10-02 00:29:10]`. One has microseconds precision, the other one does not.

I am also not sure if we should use the `precision_for_unit/1` conversion. For example, if you add microseconds to a naive date time with seconds precision, does it now have microseconds precision? Similarly, if you have a naive date time with microseconds precision and you add seconds to it, does it have seconds precision? I believe the correct approach would be to keep the precision that is already in the NaiveDateTime.